fraud/service.go

```diff
@@ -0,0 +1,192 @@
+	log.Info("fraud service is starting...")
+
+	err := f.RegisterUnmarshaler(BadEncoding, UnmarshalBEFP)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *service) Stop(context.Context) error {
+	log.Info("fraud service is stopping...")
+	return nil
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.RLock()
+	u, ok := f.unmarshalers[proofType]
+	f.mu.RUnlock()
+	if !ok {
+		return nil, errors.New("unmarshaler is not registered")
+	}
+
+	t, wasjoined, err := f.topics.getTopic(proofType)
+	if err != nil {
+		return nil, err
+	}
+	// if topic was joined for the first time then we should register a validator for it
+	if !wasjoined {
+		// add internal validation to topic inside libp2p for provided ProofType
+		if err = f.topics.pubsub.RegisterTopicValidator(
+			getSubTopic(proofType),
+			func(ctx context.Context, _ peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+				return f.processIncoming(ctx, proofType, msg.Data)
+			},
+		); err != nil {
+			return nil, err
+		}
+	}
+
+	return newSubscription(t, u)
+}
+
+func (f *service) RegisterUnmarshaler(proofType ProofType, u ProofUnmarshaler) error {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	if _, ok := f.unmarshalers[proofType]; ok {
+		return errors.New("unmarshaler is registered")
+	}
+	f.unmarshalers[proofType] = u
+
+	return nil
+}
+
+func (f *service) UnregisterUnmarshaler(proofType ProofType) error {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	if _, ok := f.unmarshalers[proofType]; !ok {
+		return errors.New("unmarshaler is not registered")
+	}
+	delete(f.unmarshalers, proofType)
+
+	return nil
+
+}
+
+func (f *service) Broadcast(ctx context.Context, p Proof) error {
+	bin, err := p.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	return f.topics.publish(ctx, bin, p.Type())
+}
+
+func (f *service) processIncoming(ctx context.Context, proofType ProofType, data []byte) pubsub.ValidationResult {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	unmarshaler, ok := f.unmarshalers[proofType]
+	if !ok {
+		log.Error("unmarshaler is not found")
+		return pubsub.ValidationReject
+	}
+	proof, err := unmarshaler(data)
+	if err != nil {
+		log.Errorw("unmarshalling header error", err)
+		return pubsub.ValidationReject
+	}
+
+	extHeader, err := f.getter(ctx, proof.Height())
+	if err != nil {
+		log.Errorw("failed to fetch block: ",
+			"err", err)
+		return pubsub.ValidationReject
+	}
+
+	err = proof.Validate(extHeader)
+	if err != nil {
+		log.Errorw("validation err: ",
+			"err", err)
+		return pubsub.ValidationReject
+	}
+	log.Debugw("received Bad Encoding Fraud Proof from block ", "hash", hex.EncodeToString(extHeader.DAH.Hash()))
+	return pubsub.ValidationAccept
+}
+
+func getSubTopic(p ProofType) string {
+	switch p {
+	case BadEncoding:
+		return "befp-sub"
+	default:
+		panic(fmt.Sprintf("invalid proof type %d", p))
+	}
+}
```

Comment:
This requires us to add a case for each new added type. Instead, you can get a topic name by p.Type() or p.Type() + "Sub"

Reply:
done